File: packages/ruby/lib/crumbtrail/body.rb
require 'json'

module Crumbtrail
  # Structured body policy. Ruby, Go and the ASP.NET Core package implement the same rules and
  # are driven by the same corpus in `test-fixtures/backend-body/cases.json`, so a divergence
  # between them fails a test instead of quietly producing three different bodies.
  module Body
    LIMIT = 16_384
    POLICY = 'crumbtrail.backend-redaction.v1'
    MAX_NESTING = 8
    MAX_KEYS = 64
    MAX_ITEMS = 40
    # Well below a phone number (10), a national insurance number (9) and a card (13 to 19).
    MAX_INTEGER_DIGITS = 6
    SAFE_INTEGER = 9_007_199_254_740_991
    DENIED = /password|passwd|passphrase|passcode|secret|token|auth|card|cvv|cvc|ssn|email|phone|address|iban|account|birth|credential|creds|cookie|session|privatekey|apikey|accesskey|securitycode|verificationcode|connection|routingnumber|taxid|nationalid|sortcode|name|postal|payload|beforejson|afterjson|mobile|contact|diagnosis|medical|patient|prescription|gender|ethnic|religion|salary|income|identifier|username|passport|insurance|beneficiary|guardian|occupation|citizen|latitude|longitude|coordinate|geolocation|province|country|street/i
    # Short words that appear inside innocent identifiers ("capacity" contains "city"), so they
    # are matched as whole words rather than as substrings.
    DENIED_WORD = /\A(?:pwd|pin|pan|otp|pass|sid|dob|zip|jwt|mfa|csrf|xsrf|city|town|geo|cell|race|sex|age|location|lat|lng|lon|gps)s?[0-9]*\z/i
    SAFE_STRING = /\A(?:[a-z][a-z_]{0,22}|[A-Z]{3}|[0-9]{1,#{MAX_INTEGER_DIGITS}})\z/
    TOKEN_STRING = /(?:sk|pk|rk|ghp|gho|ghu|ghs|glpat|xox[baprs])[-_][a-zA-Z0-9_.=-]{12,}/i
    FIELD_NAME = /\A[a-zA-Z_][a-zA-Z0-9_]*\z/
    REDACTED = '[REDACTED]'.freeze

    class Invalid < StandardError; end

    def self.capture(bytes, truncated = false)
      return [nil, 'truncated'] if truncated || bytes.bytesize > LIMIT
      return [nil, 'missing'] if bytes.empty?
      removed = [false]
      value = JSON.parse(bytes, max_nesting: MAX_NESTING, allow_duplicate_key: false)
      value = walk(value, '', removed)
      encoded = JSON.generate(value)
      return [nil, 'truncated'] if encoded.bytesize > LIMIT
      [encoded, removed[0] ? 'redacted' : 'captured']
    rescue JSON::ParserError, JSON::GeneratorError, JSON::NestingError, Invalid, EncodingError
      [nil, 'invalid']
    end

    # A sensitive key is redacted whole. The subtree under it is never walked, so a malformed
    # or oversized value inside a secret cannot turn the whole body into `invalid` and cannot
    # be inspected on the way to being dropped.
    def self.walk(value, key, removed)
      unless sensitive?(key)
        case value
        when Hash
          raise Invalid if value.size > MAX_KEYS
          return value.to_h do |k, v|
            raise Invalid unless k.bytesize <= 64 && k.match?(FIELD_NAME)
            [k, walk(v, k, removed)]
          end
        when Array
          raise Invalid if value.size > MAX_ITEMS
          return value.map { |v| walk(v, key, removed) }
        when Numeric
          return value if number?(value)
        when TrueClass, FalseClass, NilClass
          return value
        when String
          return value if value.match?(SAFE_STRING) && !value.match?(TOKEN_STRING)
        end
      end
      removed[0] = true
      REDACTED
    end

    def self.sensitive?(key)
      words = key.gsub(/([a-z0-9])([A-Z])/, '\1 \2').split(/[^a-zA-Z0-9]+/)
      key.gsub(/[^a-zA-Z0-9]/, '').match?(DENIED) || words.any? { |word| word.match?(DENIED_WORD) }
    end

    # The integer digit cap already excludes every card length. Luhn stays because it also
    # catches a card smuggled across a decimal point, where the integer part is short.
    def self.number?(value)
      return false unless value.finite?
      return false if value.abs > SAFE_INTEGER
      return false if value.to_i.abs.to_s.length > MAX_INTEGER_DIGITS
      !luhn?(digits(value))
    end

    def self.digits(value)
      value.abs.to_s.gsub(/[^0-9]/, '')
    end

    def self.luhn?(digits)
      return false unless digits.length.between?(13, 19)
      sum = 0
      twice = false
      digits.reverse.each_char do |char|
        n = char.ord - 48
        if twice
          n *= 2
          n -= 9 if n > 9
        end
        sum += n
        twice = !twice
      end
      (sum % 10).zero?
    end

    def self.json?(type)
      type.to_s.split(';', 2).first.to_s.strip.match?(%r{\Aapplication/(?:json|[^/;\s]+\+json)\z}i)
    end

    def self.metadata(field, state)
      { policy: POLICY, fields: state == 'redacted' ? [{ path: field, reason: 'backend_structured_profile', action: 'redacted' }] : [] }
    end
  end
end

File: packages/go/body.go
package crumbtrail

import (
	"bytes"
	"encoding/json"
	"io"
	"math"
	"mime"
	"regexp"
	"strconv"
	"strings"
	"unicode/utf8"
)

// Structured body policy. Go, Ruby and the ASP.NET Core package implement the same rules and
// are driven by the same corpus in test-fixtures/backend-body/cases.json, so a divergence
// between them fails a test instead of quietly producing three different bodies.
const bodyLimit = 16384
const policy = "crumbtrail.backend-redaction.v1"
const maxNesting = 8
const maxKeys = 64
const maxItems = 40

// Well below a phone number (10), a national insurance number (9) and a card (13 to 19).
const maxIntegerDigits = 6
const safeInteger = 9007199254740991

var denied = regexp.MustCompile(`(?i)password|passwd|passphrase|passcode|secret|token|auth|card|cvv|cvc|ssn|email|phone|address|iban|account|birth|credential|creds|cookie|session|privatekey|apikey|accesskey|securitycode|verificationcode|connection|routingnumber|taxid|nationalid|sortcode|name|postal|payload|beforejson|afterjson|mobile|contact|diagnosis|medical|patient|prescription|gender|ethnic|religion|salary|income|identifier|username|passport|insurance|beneficiary|guardian|occupation|citizen|latitude|longitude|coordinate|geolocation|province|country|street`)

// Short words that appear inside innocent identifiers ("capacity" contains "city"), so they are
// matched as whole words rather than as substrings.
var deniedWord = regexp.MustCompile(`(?i)^(pwd|pin|pan|otp|pass|sid|dob|zip|jwt|mfa|csrf|xsrf|city|town|geo|cell|race|sex|age|location|lat|lng|lon|gps)s?[0-9]*$`)
var camel = regexp.MustCompile(`([a-z0-9])([A-Z])`)
var nonWord = regexp.MustCompile(`[^a-zA-Z0-9]+`)
var fieldName = regexp.MustCompile(`^[a-zA-Z_][a-zA-Z0-9_]*$`)
var safeString = regexp.MustCompile(`^([a-z][a-z_]{0,22}|[A-Z]{3}|[0-9]{1,6})$`)
var tokenString = regexp.MustCompile(`(?i)(sk|pk|rk|ghp|gho|ghu|ghs|glpat|xox[baprs])[-_][a-zA-Z0-9_.=-]{12,}`)

type capturedBody struct {
	Body  any
	State string
}

func captureBody(raw []byte, truncated bool) capturedBody {
	if truncated || len(raw) > bodyLimit {
		return capturedBody{nil, "truncated"}
	}
	if len(raw) == 0 {
		return capturedBody{nil, "missing"}
	}
	if !utf8.Valid(raw) {
		return capturedBody{nil, "invalid"}
	}
	decoder := json.NewDecoder(bytes.NewReader(raw))
	decoder.UseNumber()
	removed := false
	value, ok := parseValue(decoder, "", 0, &removed)
	if !ok {
		return capturedBody{nil, "invalid"}
	}
	if _, err := decoder.Token(); err != io.EOF {
		return capturedBody{nil, "invalid"}
	}
	result, err := json.Marshal(value)
	if err != nil {
		return capturedBody{nil, "invalid"}
	}
	if len(result) > bodyLimit {
		return capturedBody{nil, "truncated"}
	}
	state := "captured"
	if removed {
		state = "redacted"
	}
	return capturedBody{string(result), state}
}

func sensitiveKey(key string) bool {
	if denied.MatchString(nonWord.ReplaceAllString(key, "")) {
		return true
	}
	for _, word := range nonWord.Split(camel.ReplaceAllString(key, "${1} ${2}"), -1) {
		if deniedWord.MatchString(word) {
			return true
		}
	}
	return false
}

func parseValue(d *json.Decoder, key string, depth int, removed *bool) (any, bool) {
	// A sensitive key is redacted whole. The subtree under it is consumed but never inspected,
	// so an oversized or oddly named field inside a secret cannot turn the whole body into
	// `invalid`, and its shape is not reported either. Ruby and .NET behave the same way.
	if sensitiveKey(key) {
		if !skipValue(d, depth) {
			return nil, false
		}
		*removed = true
		return "[REDACTED]", true
	}
	token, err := d.Token()
	if err != nil {
		return nil, false
	}
	var value any
	switch token := token.(type) {
	case json.Delim:
		if depth >= maxNesting {
			return nil, false
		}
		switch token {
		case '{':
			object := map[string]any{}
			for d.More() {
				t, err := d.Token()
				k, valid := t.(string)
				if err != nil || !valid || len(k) > 64 || !fieldName.MatchString(k) || len(object) >= maxKeys {
					return nil, false
				}
				if _, exists := object[k]; exists {
					return nil, false
				}
				v, ok := parseValue(d, k, depth+1, removed)
				if !ok {
					return nil, false
				}
				object[k] = v
			}
			if t, err := d.Token(); err != nil || t != json.Delim('}') {
				return nil, false
			}
			value = object
		case '[':
			list := []any{}
			for d.More() {
				if len(list) >= maxItems {
					return nil, false
				}
				v, ok := parseValue(d, key, depth+1, removed)
				if !ok {
					return nil, false
				}
				list = append(list, v)
			}
			if t, err := d.Token(); err != nil || t != json.Delim(']') {
				return nil, false
			}
			value = list
		default:
			return nil, false
		}
	case json.Number:
		if !safeNumber(token) {
			*removed = true
			return "[REDACTED]", true
		}
		value = token
	case string:
		if !safeString.MatchString(token) || tokenString.MatchString(token) {
			*removed = true
			return "[REDACTED]", true
		}
		value = token
	case nil, bool:
		value = token
	default:
		return nil, false
	}
	return value, true
}

// skipValue consumes one value without applying any shape rule to it. Well formedness,
// nesting depth and duplicate keys still hold, because those are properties of the document
// rather than of the redacted value.
func skipValue(d *json.Decoder, depth int) bool {
	token, err := d.Token()
	if err != nil {
		return false
	}
	delim, isDelim := token.(json.Delim)
	if !isDelim {
		return true
	}
	if depth >= maxNesting {
		return false
	}
	switch delim {
	case '{':
		seen := map[string]struct{}{}
		for d.More() {
			t, err := d.Token()
			k, valid := t.(string)
			if err != nil || !valid {
				return false
			}
			if _, exists := seen[k]; exists {
				return false
			}
			seen[k] = struct{}{}
			if !skipValue(d, depth+1) {
				return false
			}
		}
	case '[':
		for d.More() {
			if !skipValue(d, depth+1) {
				return false
			}
		}
	default:
		return false
	}
	if _, err := d.Token(); err != nil {
		return false
	}
	return true
}

// The integer digit cap already excludes every card length. Luhn stays because it also catches
// a card smuggled across a decimal point, where the integer part is short.
func safeNumber(token json.Number) bool {
	f, err := token.Float64()
	if err != nil || math.IsInf(f, 0) || math.IsNaN(f) || math.Abs(f) > safeInteger {
		return false
	}
	text := strconv.FormatFloat(math.Abs(f), 'f', -1, 64)
	integer := text
	if dot := strings.IndexByte(text, '.'); dot >= 0 {
		integer = text[:dot]
	}
	if len(integer) > maxIntegerDigits {
		return false
	}
	return !luhn(strings.Replace(text, ".", "", 1))
}

func luhn(digits string) bool {
	if len(digits) < 13 || len(digits) > 19 {
		return false
	}
	sum, twice := 0, false
	for i := len(digits) - 1; i >= 0; i-- {
		if digits[i] < '0' || digits[i] > '9' {
			return false
		}
		n := int(digits[i] - '0')
		if twice {
			n *= 2
			if n > 9 {
				n -= 9
			}
		}
		sum += n
		twice = !twice
	}
	return sum%10 == 0
}

func isJSON(contentType string) bool {
	media, _, err := mime.ParseMediaType(contentType)
	if err != nil {
		return false
	}
	return media == "application/json" || (strings.HasPrefix(media, "application/") && strings.HasSuffix(media, "+json"))
}

func redaction(field, state string) map[string]any {
	fields := []any{}
	if state == "redacted" {
		fields = append(fields, map[string]any{"path": field, "reason": "backend_structured_profile", "action": "redacted"})
	}
	return map[string]any{"policy": policy, "fields": fields}
}
